src/main/java/ch/njol/skript/expressions/arithmetic/ExprArithmetic.java

```diff
@@ -48,147 +49,190 @@
-		first = (Expression<? extends Number>) exprs[0];
-		second = (Expression<? extends Number>) exprs[1];
-		
+		first = LiteralUtils.defendExpression(exprs[0]);
+		second = LiteralUtils.defendExpression(exprs[1]);
+
+		if (!LiteralUtils.canInitSafely(first, second))
+			return false;
+
+		returnType = first.getReturnType();
+		Class<?> secondClass = second.getReturnType();
+
 		PatternInfo patternInfo = patterns.getInfo(matchedPattern);
 		op = patternInfo.operator;
-		
-		if (op == Operator.DIV || op == Operator.EXP) {
-			returnType = Double.class;
-		} else {
-			Class<?> firstReturnType = first.getReturnType();
-			Class<?> secondReturnType = second.getReturnType();
-			
-			boolean firstIsInt = false;
-			boolean secondIsInt = false;
-			for (final Class<?> i : INTEGER_CLASSES) {
-				firstIsInt |= i.isAssignableFrom(firstReturnType);
-				secondIsInt |= i.isAssignableFrom(secondReturnType);
+		leftGrouped = patternInfo.leftGrouped;
+		rightGrouped = patternInfo.rightGrouped;
+
+		if (Number.class.isAssignableFrom(returnType)) {
+			if (op == Operator.DIV || op == Operator.EXP) {
+				returnType = Double.class;
+			} else {
+				Class<?> firstReturnType = first.getReturnType();
+				Class<?> secondReturnType = second.getReturnType();
+
+				boolean firstIsInt = false;
+				boolean secondIsInt = false;
+				for (final Class<?> i : INTEGER_CLASSES) {
+					firstIsInt |= i.isAssignableFrom(firstReturnType);
+					secondIsInt |= i.isAssignableFrom(secondReturnType);
+				}
+
+				returnType = firstIsInt && secondIsInt ? Long.class : Double.class;
 			}
-			
-			returnType = firstIsInt && secondIsInt ? Long.class : Double.class;
 		}
-		
+
+		Arithmetic<?> arithmetic = null;
+
+		if (!returnType.equals(Object.class)
+				&& !secondClass.equals(Object.class)
+				&& (arithmetic = getArithmetic(returnType, op, secondClass)) == null) {
+			return false;
+		}
+
 		// Chaining
-		if (first instanceof ExprArithmetic && !patternInfo.leftGrouped) {
+		if (first instanceof ExprArithmetic && !leftGrouped) {
 			chain.addAll(((ExprArithmetic) first).chain);
 		} else {
 			chain.add(first);
 		}
 		chain.add(op);
-		if (second instanceof ExprArithmetic && !patternInfo.rightGrouped) {
+		if (second instanceof ExprArithmetic && !rightGrouped) {
 			chain.addAll(((ExprArithmetic) second).chain);
 		} else {
 			chain.add(second);
 		}
-		
-		arithmeticGettable = ArithmeticChain.parse(chain);
-		
+
+		if (arithmetic != null)
+			arithmeticGettable = ArithmeticChain.parse(chain, arithmetic);
+
 		return true;
 	}
-	
-	@SuppressWarnings("null")
+
 	@Override
-	protected Number[] get(final Event e) {
-		Number[] one = (Number[]) Array.newInstance(returnType, 1);
-		
-		one[0] = arithmeticGettable.get(e, returnType == Long.class);
-		
+	@SuppressWarnings("null")
+	protected Object[] get(final Event e) {
+		if (arithmeticGettable == null) {
```

Comment:
I'm not sure it's safe for us to assume the parsed arithmetic gettable will remain the same during other runtimes

Thus, if it's null, it should be reset back to null after get is done

Reply:
Ah, you're right